plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,420 @@
+	tags["band"] = plugin.wlanBandFromInfo(info)
+	fields := make(map[string]interface{})
+	fields["total_associations"] = totalAssociations.NewTotalAssociations
+	acc.AddGauge("fritzbox_wlan", fields, tags)
+	return nil
+}
+
+func (plugin *Fritzbox) wlanBandFromInfo(info *wlanconfig.GetInfoResponse) string {
+	band := info.NewX_AVM_DE_FrequencyBand
+	if band != "" {
+		return band
+	}
+	if 1 <= info.NewChannel && info.NewChannel <= 14 {
+		return "2400"
+	}
+	return "5000"
+}
+
+func (plugin *Fritzbox) gatherHostsInfo(acc telegraf.Accumulator, deviceClient *tr064.Client, service tr064.ServiceDescriptor) error {
+	serviceClient := hosts.ServiceClient{
+		TR064Client: deviceClient,
+		Service:     service,
+	}
+	connections, err := plugin.fetchHostsConnections(&serviceClient)
+	if err != nil {
+		return err
+	}
+	for _, connection := range connections {
+		_, err = uuid.Parse(connection.RightDeviceName)
+		if err == nil {
+			continue
+		}
+		tags := make(map[string]string)
+		tags["source"] = serviceClient.TR064Client.DeviceUrl.Hostname()
+		tags["service"] = serviceClient.Service.ShortId()
+		tags["node"] = connection.RightDeviceName
+		tags["node_role"] = plugin.hostRole(connection.RightMeshRole)
+		tags["node_ap"] = connection.LeftDeviceName
+		tags["node_ap_role"] = plugin.hostRole(connection.LeftMeshRole)
+		tags["link_type"] = connection.InterfaceType
+		tags["link_name"] = connection.InterfaceName
+		fields := make(map[string]interface{})
+		fields["max_data_rate_tx"] = connection.MaxDataRateTx
+		fields["max_data_rate_rx"] = connection.MaxDataRateRx
+		fields["cur_data_rate_tx"] = connection.CurDataRateTx
+		fields["cur_data_rate_rx"] = connection.CurDataRateRx
+		acc.AddGauge("fritzbox_hosts", fields, tags)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) hostRole(role string) string {
+	if role == "unknown" {
+		return "client"
+	}
+	return role
+}
+
+func (plugin *Fritzbox) fetchHostsConnections(serviceClient *hosts.ServiceClient) ([]*mesh.Connection, error) {
+	meshListPath := &hosts.X_AVM_DE_GetMeshListPathResponse{}
+	err := serviceClient.X_AVM_DE_GetMeshListPath(meshListPath)
+	if err != nil {
+		return nil, err
+	}
+	meshListResponse, err := serviceClient.TR064Client.Get(meshListPath.NewX_AVM_DE_MeshListPath)
+	if err != nil {
+		return nil, err
+	}
+	if meshListResponse.StatusCode == http.StatusNotFound {
+		return make([]*mesh.Connection, 0), nil
+	}
+	if meshListResponse.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to fetch mesh list %q (status: %s)", meshListPath.NewX_AVM_DE_MeshListPath, meshListResponse.Status)
+	}
+	defer meshListResponse.Body.Close()
+	meshListBytes, err := io.ReadAll(meshListResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+	meshList := &mesh.List{}
+	err = json.Unmarshal(meshListBytes, meshList)
+	if err != nil {
+		return nil, err
+	}
+	return meshList.Connections(), nil
+}
+
+func init() {
+	inputs.Add(pluginName, func() telegraf.Input {
+		return defaultFritzbox()
```

Comment:
Please directly create the struct here instead of using a function.

Reply:
Adapted via 1ada7a8c36211e8f74afa86ed9638c45083686bb